CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,475 @@
+```typescript
+// POST to /getBlindedMessagePartialSig
+// Headers:
+//   Content-Type: application/json
+//   Authorization: ${signature from authorized signer (e.g. DEK) for account}
+interface GetBlindedMessageSigRequest {
+  // Celo account address. Query is charge against this account's quota.
+  account: string
+  // Query message. A blinded elliptic curve point encoded in base64.
+  blindedQueryPhoneNumber: string
+  // Optional on-chain identifier. Unlocks additional quota if the
+  // account is verified as an owner of the identifier.
+  hashedPhoneNumber?: string
+  // [Deprecated] Client-specified request timestamp, in milliseconds.
+  // Used for request expiration.
+  timestamp?: number
+  // Client-specified session ID.
+  sessionID?: string
+}
+```
+
+#### Extension for domain-restricted requests
+
+In order to support domain-restricted queries, a new API is proposed:
+
+```typescript
+// POST to /getDomainRestrictedPartialSig
+// Headers:
+//   Content-Type: application/json
+interface GetDomainRestrictedSigRequest {
+  // Domain specification.
+  // Selects the PRF domain and rate limiting rules.
+  domain: Domain
+  // Domain-specific options.
+  // Used for inputs relevant to the domain, but not part of the domain string.
+  // Example: { "authorization": <signature> } for an account-restricted domain.
+  domainOptions?: DomainOptions
+  // Query message. A blinded elliptic curve point encoded in base64.
+  blindedMessage: string
+  // Client-specified session ID.
+  sessionID?: string
+}
+
+interface Domain {
+  // Unique name of the domain. (e.g. "ODIS Password Domain")
+  name: string
+  // Major version number. Allows for backwards incompatible changes.
+  version: number
+  // Arbitrary key-value pairs.
+  // MUST be serializable to EIP-712 encoding.
+  [key: string]: EIP712Value
+}
+
+interface DomainOptions {
+  // Arbitrary key-value pairs.
+  // Valid keys and values depend on the domain specification.
+  // SHOULD be serializable to EIP-712 encoding.
+  [key: string]: EIP712Value
+}
+```
+
+### Domain specifiers
+
+Domain specifiers serve two functions:
+
+- Define the input domain, such that it is uncorrelated to the output in any other domain.
+- Specify the rate limiting function to be used in deciding whether to serve the request.
+
+In order to provide flexibility and support future extensions, the domain specifier is structured
+and allows for the usage of multiple domain-specific rate limiting rules, as support by the ODIS
+service.
+
+In the domain specifiers, two fields MUST be provided: `name` and `version`. Together, these select
+which set of rules the service will apply to the request. Rule sets are included in the ODIS
+implementation, and new rule sets may be added over time. Updating a rule set in a way that would
+change the rules applied to existing domains SHOULD be accompanied by an increase in the `version`
+number. Domain names SHOULD be human-readable. Version numbers SHOULD be whole integers.
```

Comment:
Does `name` correspond to a specific type of identity/input (e.g. phone number, password) which the protocol can then look up the corresponding rule set/rate-limiting scheme for? Or should the `name` explicitly be the rate-limiting scheme (e.g. `Linear Backoff Domain`)?

Reply:
I try to make this clear above, but the `name` identifies a ruleset, so it could be `Linear Backoff Domain` or `Password Domain`, but it cannot include a specific user identifier such as a phone number. (i.e. it must come from a small fixed set, included in the ODIS software). I don't specify whether the name should refer to what the ruleset does (e.g. linear backoff), or how its intended to be used (e.g. to hash passwords). I think its ok to do either and both, but I lean toward the former.